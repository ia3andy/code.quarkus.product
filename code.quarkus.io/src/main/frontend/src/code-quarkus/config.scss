@import url(https://fonts.googleapis.com/css?family=PT+Mono);
@import "../red-hat-font.scss";
@include printRedHatFont(400, "Regular", $familyName: "RedHatText", $relative: false);
@include printRedHatFont(700, "Medium", $familyName: "RedHatText", $relative: false);

$font: RedHatText, Regular;
$inputFont: "PT Mono", Regular;

$quarkusBlue: #4B97E8;

$background: white;
$backgroundSecondary: #f0f0f0;

$primary: #646464;
$secondary: #EB0B19;

$warningColor: #ff004a;
$errorColor: #ff0048a8;

$linkTextColor: $secondary;

$readOnlyTextColor: #bdbcbc;
$readonlyLabelAndBorderColor: #707070;

$loaderColor: $primary;

// Top container
$topContainerTextColor: black;
$topContainerBg: $backgroundSecondary;
$topContainerBgImage: none;

$fieldLabelAndBorderColor: $primary; 
$fieldLabelAndBorderHoverColor: $secondary;
$fieldTextColor: $topContainerTextColor; 
$selectOptionsTextColor: black; 

$readOnlyTextColor: #a8a7a7;
$readonlyLabelAndBorderColor: #999999;

$generateButtonTextColor: white;
$generateButtonBgColor: $secondary;
$generateButtonBorderColorOnHover: black;
$generateButtonDisabledBgAndBorderColor: rgba(90, 149, 235, 0.342);
$generateButtonDisabledColor: rgba(255, 255, 255, 0.384);

$generateButtonMoreDropdownTextColor: black;
$generateButtonMoreDropdownBg: #d4d4d4;

// Main container
$mainContainerTextColor: black;
$mainContainerBg: $background;

$mobileWidth: 900px;
$tabletWidth: 1200px;

// Next steps Modal
$modalBg: $backgroundSecondary;
$modalTextColor: black;
$modalBorderColor: $secondary;
$modalTitleColor: $secondary;
$modalLinkColor: $linkTextColor;

$modalCloseButtonBg: $primary;
$modalCloseButtonBgHover: $secondary;

$nextStepsModalDownloadButtonBg: $secondary;
$nextStepsModalDownloadButtonTextColor: white;
$nextStepsModalDownloadButtonBorderColorOnHover: black;

$modalFooterButtonBg: white;
$modalFooterButtonTextColor: $secondary;

$codeBg: #2b2b2b;
$codeTextColor: white;

// Extensions picker
$extensionsPickerTextColor: $mainContainerTextColor;
$extensionsPickerCheckColor: $mainContainerTextColor;
$extensionsPickerCheckSelectedHoverColor: $mainContainerTextColor;
$extensionsPickerCheckSelectedColor: #0F69C9;
$extensionsPickerCheckSelectedByDefaultColor: $readonlyLabelAndBorderColor;
$extensionsPickerKeyboardActivedTextColor: #0F69C9;
$extensionsPickerRemoveButtonColor: $warningColor;

$extensionsPickerDescriptionTextColorOnTablet: #636363;

$extensionsPickerMoreDropdownBg: $backgroundSecondary;
$extensionsPickerMoreDropdownTextColor: black;

$extensionsPickerCategoryTextColor: $mainContainerTextColor;
$extensionsPickerCategoryUnderlineColor: #dcdcdc;

$extensionsPickerSearchTextColor: $mainContainerTextColor;
$extensionsPickerSearchBorderColor: #dcdcdc;
$extensionsPickerSearchBorderColorOnFocus: $secondary;
$extensionsPickerSearchClearColor: $warningColor;

$extensionsPickerFlagExampleColor: $primary;
$extensionsPickerFlagPreviewColor: $secondary;
$extensionsPickerFlagExperimentalColor: $warningColor;
$extensionsPickerFlagDeprecatedColor: $readonlyLabelAndBorderColor;
$extensionsPickerFlagCustomColor: black;

// Header
$headerBackgroundColor: #1A1A1A;
$headerBackgroundImage: none;
$headerTextColor: white;

$headerMenuLinkTextColor: $quarkusBlue;
$headerQuarkusVersionTextColor: white;
$headerQuarkusVersionBorderColor: $primary;

$headerMenuStartCodingButtonTextColor: white;
$headerMenuStartCodingButtonBorderColor: $secondary;
$headerMenuStartCodingButtonBg: $secondary;

$headerSupportButtonTextColor: white;
$headerSupportButtonBorderColor: white;
$headerSupportButtonBackgroundColor: transparent;

$headerSupportPanelBackgroundColor: black;
$headerSupportPanelTextColor: white;

// Blurb
$blurbBg: $backgroundSecondary;
$blurbTextColor: black;
$blurbLinkTextColor: $linkTextColor;
$blurbInfoIconColor: $primary;