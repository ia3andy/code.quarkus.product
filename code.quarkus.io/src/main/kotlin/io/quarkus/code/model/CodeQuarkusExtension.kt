package io.quarkus.code.model

data class CodeQuarkusExtension(
        val id: String,
        val shortId: String,
        val version: String,
        val name: String,
        val description: String?,
        val shortName: String?,
        val category: String,
        val tags: List<String>,
        val keywords: List<String>,
        val providesExampleCode: Boolean,
        val guide: String?,
        val order: Int,

        @Deprecated(message = "no continued")
        val default: Boolean,

        @Deprecated(message = "has been replaced", replaceWith = ReplaceWith("tags"))
        val status: String,

        @Deprecated(message = "has been replaced", replaceWith = ReplaceWith("keywords"))
        val labels: List<String>
)