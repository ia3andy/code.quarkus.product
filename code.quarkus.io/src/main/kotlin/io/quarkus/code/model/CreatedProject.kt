package io.quarkus.code.model

data class CreatedProject(
        var path: String
)