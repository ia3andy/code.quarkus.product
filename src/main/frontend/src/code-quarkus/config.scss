@import url(https://fonts.googleapis.com/css?family=PT+Mono);
@import url(https://fonts.googleapis.com/css?family=Open+Sans);

$font: "Open Sans", Regular;
$inputFont: "PT Mono", Regular;

$background: #0d1c2c;
$backgroundSecondary: black;

$primary: #71aeef;
$secondary: #4695eb;

$warningColor: #ff004a;
$errorColor: #ff0048a8;

$linkTextColor: $secondary;

$readOnlyTextColor: #bdbcbc;
$readonlyLabelAndBorderColor: #707070;

$loaderColor: white;

// Top container
$topContainerTextColor: white;
$topContainerBg: $background;
$topContainerBgImage: url("/static/media/bg-form.jpg");

$fieldLabelAndBorderColor: $primary; 
$fieldLabelAndBorderHoverColor: $topContainerTextColor; 
$fieldTextColor: $topContainerTextColor; 
$selectOptionsTextColor: black; 

$readOnlyTextColor: #bdbcbc;
$readonlyLabelAndBorderColor: #707070;

$generateButtonTextColor: white;
$generateButtonBgColor: $secondary;
$generateButtonBorderColorOnHover: white;
$generateButtonDisabledBgAndBorderColor: rgba(90, 149, 235, 0.342);
$generateButtonDisabledColor: rgba(255, 255, 255, 0.384);

$generateButtonMoreDropdownTextColor: white;
$generateButtonMoreDropdownBg: black;

// Main container
$mainContainerTextColor: white;
$mainContainerBg: $background;

$mobileWidth: 900px;
$tabletWidth: 1200px;

// Next steps Modal
$modalBg: $background;
$modalTextColor: white;
$modalBorderColor: $secondary;
$modalTitleColor: $secondary;
$modalLinkColor: $linkTextColor;

$modalCloseButtonBg: $secondary;
$modalCloseButtonBgHover: rgba(70, 149, 235, 0.548);

$nextStepsModalDownloadButtonBg: $secondary;
$nextStepsModalDownloadButtonTextColor: white;
$nextStepsModalDownloadButtonBorderColorOnHover: white;

$modalFooterButtonBg: black;
$modalFooterButtonTextColor: $secondary;

$codeBg: black;
$codeTextColor: white;

// Extensions picker
$extensionsPickerTextColor: $mainContainerTextColor;
$extensionsPickerCheckColor: white;
$extensionsPickerCheckSelectedHoverColor: white;
$extensionsPickerCheckSelectedColor: $primary;
$extensionsPickerCheckSelectedByDefaultColor: $readonlyLabelAndBorderColor;
$extensionsPickerKeyboardActivedTextColor: $secondary;
$extensionsPickerRemoveButtonColor: $warningColor;

$extensionsPickerDescriptionTextColorOnTablet: #ccc;

$extensionsPickerMoreDropdownBg: $backgroundSecondary;
$extensionsPickerMoreDropdownTextColor: white;

$extensionsPickerCategoryTextColor: $mainContainerTextColor;
$extensionsPickerCategoryUnderlineColor: $secondary;

$extensionsPickerSearchTextColor: $mainContainerTextColor;
$extensionsPickerSearchBorderColor: $extensionsPickerSearchTextColor;
$extensionsPickerSearchBorderColorOnFocus: $secondary;
$extensionsPickerSearchClearColor: $warningColor;

$extensionsPickerFlagExampleColor: $primary;
$extensionsPickerFlagPreviewColor: $secondary;
$extensionsPickerFlagExperimentalColor: $warningColor;
$extensionsPickerFlagCustomColor: black;

// Header
$headerBackgroundColor: $background;
$headerBackgroundImage: url("/static/media/bg-header.jpg");
$headerTextColor: white;

$headerQuarkusVersionTextColor: white;
$headerQuarkusVersionBorderColor: $primary;

$headerMenuLinkTextColor: white;

$headerMenuStartCodingButtonTextColor: white;
$headerMenuStartCodingButtonBorderColor: $secondary;
$headerMenuStartCodingButtonBg: $secondary;

$headerSupportButtonTextColor: white;
$headerSupportButtonBorderColor: white;
$headerSupportButtonBackgroundColor: transparent;

$headerSupportPanelBackgroundColor: black;
$headerSupportPanelTextColor: white;

// Blurb
$blurbBg: #193E62;
$blurbTextColor: white;
$blurbLinkTextColor: $linkTextColor;
$blurbInfoIconColor: $primary;