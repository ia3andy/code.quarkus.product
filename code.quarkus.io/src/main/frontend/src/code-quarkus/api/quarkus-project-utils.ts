import { parse, stringify } from 'querystring';
import { ExtensionEntry } from '../pickers/extensions-picker';
import { createGitHubProject } from './code-quarkus-github-api';
import { QuarkusProject } from './model';
import { BACKEND_URL, CLIENT_NAME } from './env';

export enum Target {
  DOWNLOAD = 'DOWNLOAD',
  SHARE = 'SHARE',
  GITHUB = 'GITHUB'
}

export interface GenerateResult {
  target: Target;
  url: string;
  shareUrl?: string;
}

export function generateProjectQuery(project: QuarkusProject, github: boolean = false): string {
  const params: any = {
    ...(project.metadata.groupId && { g: project.metadata.groupId }),
    ...(project.metadata.artifactId && { a: project.metadata.artifactId }),
    ...(project.metadata.version && { v: project.metadata.version }),
    ...(project.metadata.buildTool && { b: project.metadata.buildTool }),
    ...(project.metadata.noExamples && { ne: project.metadata.noExamples }),
    ...(project.extensions && { s: project.extensions.map(e => e.shortId).join('.') }),
    cn: CLIENT_NAME
  };
  if (github) {
    params.github = true;
  }
  return stringify(params);
}

const BASE_LOCATION = window.location.href.replace(window.location.search, '');

export function getProjectDownloadUrl(project: QuarkusProject) {
  const baseUrl = BACKEND_URL.startsWith('http') ? BACKEND_URL : BASE_LOCATION;
  return `${baseUrl.replace(/\/$/, '')}/d?${generateProjectQuery(project)}`;
}

export function getProjectShareUrl(project: QuarkusProject, github = false) {
  return `${BASE_LOCATION}?${generateProjectQuery(project, github)}`;
}

export async function generateProject(environment: string, project: QuarkusProject, target: Target): Promise<GenerateResult> {
  switch (target) {
    case Target.DOWNLOAD:
    case Target.SHARE:
      const url = getProjectDownloadUrl(project);
      const shareUrl = getProjectShareUrl(project);
      if (target !== Target.SHARE && environment !== 'dev') {
        setTimeout(() => window.open(url, '_blank'), 500);
      }
      return { target, url, shareUrl };
    case Target.GITHUB:
      const result = await createGitHubProject(project);
      return { target, url: result.url };
  }
}

export const createOnGitHub = (project: QuarkusProject, clientId: string) => {
  const authParams = {
    redirect_uri: getProjectShareUrl(project, true),
    client_id: clientId,
    scope: 'public_repo,workflow',
    state: Math.random().toString(36)
  };
  const githubAuthorizeUrl = `https://github.com/login/oauth/authorize?${stringify(authParams)}`;
  window.location.href = githubAuthorizeUrl;
};

export function newDefaultProject(): QuarkusProject {
  return ({
    metadata: {
      groupId: 'org.acme',
      artifactId: 'code-with-quarkus',
      version: '1.0.0-SNAPSHOT',
      buildTool: 'MAVEN',
      noExamples: false
    },
    extensions: [],
  });
}

const queryName = 'extension-search';

export function resolveInitialFilterQueryParam(): string {
  return getParams(queryName) || '';
}

const getParams = (paramName: string): string | null => {
  const searchParams = new URLSearchParams(window.location.search);
  return searchParams.get(paramName);
}

export function resolveInitialProject(extensions: ExtensionEntry[]) {
  return parseProjectInQuery(extensions) || newDefaultProject();
}

const defaultCleanHistory = () => {
  console.log(`remove query from url: ${window.location.search}`);
  window.history.replaceState({}, document.title, window.location.href.replace(window.location.search, ''));
};

export function parseProjectInQuery(extensions: ExtensionEntry[],
                                    search: string = window.location.search.substr(1),
                                    cleanHistory: () => void = defaultCleanHistory): QuarkusProject | undefined {
  if (search.length === 0) {
    return undefined;
  }
  const queryObj = parse(search);
  const shortIds = new Set((typeof queryObj.s === 'string') ? (queryObj.s as string).split('.') : []);
  const ids = new Set();
  const selectedExtensions = extensions.filter(e => {
    if (shortIds.has(e.shortId)) {
      const already = ids.has(e.id);
      ids.add(e.id);
      return !already;
    }
    return false;
  });
  const defaultProj = newDefaultProject();
  const project = {
    metadata: {
      groupId: queryObj.g || defaultProj.metadata.groupId,
      artifactId: queryObj.a || defaultProj.metadata.artifactId,
      version: queryObj.v || defaultProj.metadata.version,
      buildTool: queryObj.b || defaultProj.metadata.buildTool,
      noExamples: queryObj.ne || defaultProj.metadata.noExamples
    },
    extensions: selectedExtensions,
    github: queryObj.github === 'true' ? {
      code: queryObj.code,
      state: queryObj.state
    } : undefined
  } as QuarkusProject;
  if (project.github) {
    console.log('Received GitHub auth');
  }
  cleanHistory();
  return project;
}