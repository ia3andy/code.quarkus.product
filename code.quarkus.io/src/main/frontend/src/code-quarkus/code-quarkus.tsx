import { Analytics, AnalyticsContext, GoogleAnalytics, useAnalytics } from '../core';
import React, { useEffect, useState } from 'react';
import { Header } from './layout/header';
import './code-quarkus.scss';
import { ExtensionsLoader } from './loaders/extensions-loader';
import { QuarkusProjectFlow } from './quarkus-project/quarkus-project-flow';
import { Config } from './api/model';


export interface CodeQuarkusProps {
  config: Config;
}

export function CodeQuarkus(props: CodeQuarkusProps) {
  const [analytics, setAnalytics] = useState<Analytics>(useAnalytics());

  useEffect(() => {
    setAnalytics((prev) => {
      const newAnalytics = props.config.gaTrackingId ? new GoogleAnalytics(props.config.gaTrackingId) : prev;
      newAnalytics.init();
      return newAnalytics;
    });
  }, [props.config.gaTrackingId]);

  return (
    <AnalyticsContext.Provider value={analytics}>
      <div className="code-quarkus">
        <Header quarkusVersion={props.config.quarkusVersion} />
        <ExtensionsLoader name="extensions">
          {extensions => (
            <QuarkusProjectFlow {...props} extensions={extensions}/>
          )}
        </ExtensionsLoader>
      </div>
    </AnalyticsContext.Provider>
  );
}
