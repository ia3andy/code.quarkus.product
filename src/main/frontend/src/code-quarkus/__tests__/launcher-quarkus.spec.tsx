import { act, cleanup, fireEvent, render, RenderResult } from '@testing-library/react';
import { wait } from '@testing-library/dom';
import * as React from 'react';
import { CodeQuarkus } from '../code-quarkus';
import { Config } from '../api/model';

jest.mock('../api/code-quarkus-api', () => ({
  fetchExtensions: async () => ([
    {
      'id': 'io.quarkus:quarkus-arc',
      'shortId': '8mc',
      'version': 'test-version',
      'name': 'ArC',
      'keywords': [
        'arc',
        'cdi',
        'dependency-injection',
        'di'
      ],
      'description': 'Build time CDI dependency injection',
      'shortName': 'CDI',
      'category': 'Core',
      'order': 2
    },
    {
      'id': 'io.quarkus:quarkus-resteasy',
      'shortId': 'ogy',
      'version': 'test-version',
      'name': 'RESTEasy JAX-RS',
      'keywords': [
        'resteasy',
        'jaxrs',
        'web',
        'rest'
      ],
      'description': 'REST framework implementing JAX-RS and more',
      'shortName': 'jax-rs',
      'category': 'Web',
      'order': 4
    },
    {
      'id': 'io.quarkus:quarkus-resteasy-jsonb',
      'shortId': '14pb',
      'version': 'test-version',
      'name': 'RESTEasy JSON-B',
      'keywords': [
        'resteasy-jsonb',
        'jaxrs-json',
        'resteasy-json',
        'resteasy',
        'jaxrs',
        'json',
        'jsonb'
      ],
      'description': 'JSON-B serialization support for RESTEasy',
      'category': 'Web',
      'order': 5
    },
    {
      'id': 'io.quarkus:quarkus-resteasy-jackson',
      'shortId': '1aen',
      'version': 'test-version',
      'name': 'RESTEasy Jackson',
      'keywords': [
        'resteasy-jackson',
        'jaxrs-json',
        'resteasy-json',
        'resteasy',
        'jaxrs',
        'json',
        'jackson'
      ],
      'description': 'Jackson serialization support for RESTEasy',
      'category': 'Web',
      'order': 6
    },
  ]),
  fetchConfig: async () => {
    throw new Error('not used');
  }
}));

afterEach(() => {
  cleanup();
});

const config = { environment: 'test', quarkusVersion: 'test-version', gitCommitId: 'test-commitid' } as Config;

it('Render CodeQuarkus', async () => {
  let comp: RenderResult;
  await act(async () => {
    comp = render(<CodeQuarkus config={config}/>);
    await comp.findByLabelText('Extensions picker');
  });
});

it('Let user Generate default application', async () => {
  window.open = jest.fn();
  let comp: RenderResult;
  await act(async () => {
    comp = render(<CodeQuarkus config={config}/>);
    await comp.findByLabelText('Extensions picker');
  });

  // Generate
  const generateBtn = await comp!.findByLabelText('Generate your application');
  fireEvent.click(generateBtn);

  const downloadLink = await comp!.findByLabelText('Download the zip');
  expect(downloadLink.getAttribute('href')).toMatchSnapshot();
});

it('Let user customize an Application and Generate it', async () => {
  window.open = jest.fn();
  let comp: RenderResult;
  await act(async () => {
    comp = render(<CodeQuarkus config={config}/>);
    await comp.findByLabelText('Extensions picker');
  });

  // Custom GAV
  const groupIdInput = await comp!.findByLabelText('Edit groupId');
  fireEvent.change(groupIdInput, { target: { value: 'io.test.group' } });
  const artifactIdInput = await comp!.findByLabelText('Edit artifactId');
  fireEvent.change(artifactIdInput, { target: { value: 'custom-test-app' } });
  const toggleMoreOptionsBtn = await comp!.findByLabelText('Toggle panel');
  fireEvent.click(toggleMoreOptionsBtn);
  const versionInput = await comp!.findByLabelText('Edit project version');
  fireEvent.change(versionInput, { target: { value: '1.0.0-TEST' } });

  // Select extensions
  const ext1 = await comp!.findByLabelText('Switch io.quarkus:quarkus-arc extension');
  fireEvent.click(ext1);

  const ext2 = await comp!.findByLabelText('Switch io.quarkus:quarkus-resteasy extension');
  fireEvent.click(ext2);

  const ext3 = await comp!.findByLabelText('Switch io.quarkus:quarkus-resteasy-jackson extension');
  fireEvent.click(ext3);

  // Generate
  const generateBtn = await comp!.findByLabelText('Generate your application');
  fireEvent.click(generateBtn);

  const downloadLink = await comp!.findByLabelText('Download the zip');
  expect(downloadLink.getAttribute('href')).toMatchSnapshot();
});