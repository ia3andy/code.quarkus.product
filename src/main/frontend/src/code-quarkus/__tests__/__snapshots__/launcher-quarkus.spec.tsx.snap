// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Let user Generate default application 1`] = `"http://localhost/d?cn=localhost"`;

exports[`Let user customize an Application and Generate it 1`] = `"http://localhost/d?g=io.test.group&a=custom-test-app&v=1.0.0-TEST&s=8mc.ogy.1aen&cn=localhost"`;
